### PR TITLE
Reduce templatization of C API and refactor for InitOrtValue

### PR DESCRIPTION
**Pull request by yuslepukhin**
Simplify API

**Description**: 
Introduce Tensor::InitOrtValue for a commonly used code pattern.
Reduce templatization of C API

**Motivation and Context**
Reduce binary footprint and simplify code.

### CODE REVIEW

File: onnxruntime/core/framework/tensor.cc [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #include "core/common/safeint.h"
 #include "core/framework/allocatormgr.h"
 #include "core/framework/data_types.h"
+#include "core/framework/ml_value.h"
 #include "core/framework/utils.h"
```
Comment (skottmckay) [inline onnxruntime/core/framework/tensor.cc:11]: nit: I have a dream, that one day we rename this file to match the renamed class it provides. :)

**Review by skottmckay — DISMISSED** [review]
:shipit:

**Review by pranavsharma — DISMISSED** [review]
What was the reduction in the binary size?

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 2]
```diff
@@ -1254,46 +1109,23 @@ static ORT_STATUS_PTR OrtGetValueImplSeqOfMap(const OrtValue* p_ml_value, int in
...
-    const auto* tensor_data = tensor.Data<TensorElemType>();
-    OrtStatus* st = OrtApis::CreateTensorAsOrtValue(allocator, shape.GetDims().data(), shape.NumDimensions(),
-                                                    onnxruntime::utils::GetONNXTensorElementDataType<TensorElemType>(), out);
-    //TODO: check overflow before doing static_cast
-    return st ? st : PopulateTensorWithData(*out, tensor_data, static_cast<size_t>(shape.Size()), sizeof(TensorElemType));
-  }
-};
-
 // Return status instead of throwing if unsupported type specified
 struct UnsupportedReturnFailStatus {
```
Comment (yuslepukhin) [inline onnxruntime/core/session/onnxruntime_c_api.cc:1130]: >UnsupportedReturnFailStatus [](http://example.com/codeflow?start=7&length=27)

Consider removing

**Comment by yuslepukhin** [comment]
> What was the reduction in the binary size?

8K according to SizeBench on RelWithDebInfo build

**Review by pranavsharma — APPROVED** [review]

**Pull request merged**